asio/include/xtt/asio/context.hpp

```diff
@@ -0,0 +1,94 @@
+/******************************************************************************
+ *
+ * Copyright 2018 Xaptum, Inc.
+ * 
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ * 
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License
+ *
+ *****************************************************************************/
+
+#ifndef XTT_ASIO_CONTEXT_HPP
+#define XTT_ASIO_CONTEXT_HPP
+#pragma once
+
+#include <xtt/cpp.hpp>
+#include <boost/asio/ip/tcp.hpp>
+#include <boost/asio/io_context.hpp>
+#include <boost/asio/io_context_strand.hpp>
+#include <boost/asio/bind_executor.hpp>
+#include <boost/asio/write.hpp>
+#include <boost/asio/read.hpp>
+
+#include <memory>
+#include <unordered_map>
+#include <functional>
+
+namespace xtt {
+namespace asio {
+
+    using server_certificate_map = std::unordered_map<suite_spec, std::unique_ptr<server_certificate_context>>;
+
+    class context {
+    public:
+        context(boost::asio::ip::tcp::socket tcp_socket,
+                const server_certificate_map& cert_map,
+                server_cookie_context& cookie_ctx);
+
+        const boost::asio::ip::tcp::socket& lowest_layer() const;
+        boost::asio::ip::tcp::socket& lowest_layer();
+
+        template <typename GPKLookupCallback,
+                  typename AssignIdCallback,
+                  typename SuccessCallback,
+                  typename ErrorCallback>
+        void async_handle_connect(GPKLookupCallback gpk_lookup_callback,
+                                  AssignIdCallback assign_id_callback,
```

Comment:
Perhaps the param names for `gpk_lookup_callback` and `assign_id_callback` should be `async_gpk_lookup` and `async_assign_id` to emphasize that they should be (we're not there yet, but it's the goal) subject to the same requirements as a build in `async_X(...)` ASIO method.